File: python/day3/part1.py
import functools

def getInput(file):
    with open(file) as f:
        return f.read().splitlines()

def countTrees(lines, slope: tuple) -> int:
    width = len(lines[0])
    # current = (position, count)
    def update(current: tuple, line: str):
        position = (current[0] + slope[0]) % width
        count = current[1] + 1 if (line[position] == '#') else current[1]
        return (position, count)

    result = functools.reduce(update, lines, (0, 0))
    return result[1]

if __name__ == '__main__':
    lines = getInput('../../inputs/day3.txt')
    slope = (3,1)
    print(countTrees(lines[1:], slope))
    

File: scala/day3/part1.scala
package com.rjhallsted.adventOfCode2020.day3

import scala.io.Source

object Part1 {
    def getInputs(filepath: String): List[String] =    
        Source.fromFile(filepath).getLines().toList

    final case class Slope(
        right: Int,
        down: Int
    )

    def countTrees(lines: List[String], slope: Slope): BigInt = {
        final case class Current(
            position: Int,
            count: BigInt,
            linesToSkip: Int
        )

        val width = lines(0).length

        def update(current: Current, line: String): Current = if (current.linesToSkip - 1 == 0) {
            val position = (current.position + slope.right) % width
            val count = if (line(position) == '#') current.count + 1 else current.count
            return Current(position, count, slope.down)
        } else {
            return current.copy(linesToSkip = current.linesToSkip - 1)
        }

        lines.foldLeft[Current](Current(0,0,slope.down))(update).count
    }
}

object Main1 extends App {
    import Part1._

    val file = "../../inputs/day3.txt"
    val lines = getInputs(file)
    val slope = Slope(3,1)

    println(countTrees(lines.tail, slope))
}